### PR TITLE
Return stale data over error

### PR DESCRIPTION
**Pull request by calebmer**
This adds an initial naïve implementation that prefers stale data over partial data. When we are about to update our observers and we see that we can only get partial data from the store, instead of throwing an error (which was the old behavior) we instead return the data from the last result and set a `stale` flag to true.

Theoretically, a more detailed algorithm would only return stale data for individual objects instead of the entire result, but this should be more than enough to provide a better user experience than an error.

This is another item on our 1.0 roadmap.

- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change

### CODE REVIEW

File: test/QueryManager.ts [hunk 1]
```diff
@@ -2859,26 +2859,156 @@ describe('QueryManager', () => {
...
+        (result) => {
+          assert.deepEqual(result, {
+            data: data1,
+            loading: false,
+            networkStatus: NetworkStatus.ready,
+            stale: false,
+          });
+        },
+        (result) => {
+          assert.deepEqual(result, {
```
Comment (helfer) [inline test/QueryManager.ts]: I feel like we would almost need to add a `partial` flag for this case. However, let's see if anyone has issues with it before we add it, since we're planning on removing the `returnPartialData` option in the near future anyway.

File: src/core/QueryManager.ts [hunk 2]
```diff
@@ -384,25 +384,43 @@ export class QueryManager {
...
+              config: this.reducerConfig,
+              previousResult: lastResult && lastResult.data,
+            });
+
+            let resultFromStore: ApolloQueryResult<T>;
+
+            // If there is some data missing and the user has told us that they
+            // do not tolerate partial data then we want to return the previous
+            // result and mark it as stale.
+            if (isMissing && !(options.returnPartialData || noFetch)) {
```
Comment (calebmer) [inline src/core/QueryManager.ts]: Because it was in the last conditional for `returnPartialData`: https://github.com/apollographql/apollo-client/pull/1306/files#diff-f28988ac35502178bef737788c3d4cd5L392

This is probably because if a user is not going to fetch a query and will instead only read that query from the cache then `returnPartialData` should implicitly be turned on? Of course, this is not `options.noFetch` it appears to be something differently entirely…

Basically, it is there to not introduce a breaking change.

File: src/core/QueryManager.ts [hunk 3]
```diff
@@ -384,25 +384,43 @@ export class QueryManager {
...
+              resultFromStore = {
+                data: lastResult.data,
```
Comment (helfer) [inline src/core/QueryManager.ts]: I just realized that this could be undefined. See issue #1363.

Comment (calebmer) [inline src/core/QueryManager.ts]: In this case we should throw an error then, correct?

Comment (helfer) [inline src/core/QueryManager.ts]: Yes, I think we should. (It won't fix #1363, but the underlying problem for that is fragment matching anyway.)
